Automattic/woocommerce-payments
# Add a notification badge to disputes admin menu item indicating disputes needing response

Fixes #354 
Fixes #4124

#### Changes proposed in this Pull Request

This PR fetches the dispute summary from server and caches it using the [`Database_Cache`](https://github.com/Automattic/woocommerce-payments/blob/444e8fb10755415a43fb18b7ca2426b1ee3543be/includes/class-database-cache.php) class. 

- Using `Database_Cache` means it automatically comes with transient like behaviour (expiration functionality).
- By default the cache will last for [24 hours](https://github.com/automattic/woocommerce-payments/blob/fix/354/includes/class-database-cache.php#L303-L304) unless we specify otherwise.
- `Database_Cache` has a getter ([`get_or_add()`](https://github.com/automattic/woocommerce-payments/blob/fix/354/includes/class-database-cache.php#L59-L92)) which will attempt to get the value from the cache, check the expiration, validate the value (via a callable). It will also fetch set the value if it isn't valid or previously set.
- The validation function isn't super useful in our situation because we can't tell from the value itself if it is invalid but I have passed the `is_array` function incase the cache is corrupted.

This dispute status count is then used to display a badge to merchants on the **Payments > Disputes** admin menu item. 

<img width="190" alt="Screen Shot 2022-06-07 at 2 32 42 pm" src="https://user-images.githubusercontent.com/8490476/172296544-4b3c1062-d0d9-4e64-9dc4-9bd029590c5c.png">

This data is cached (against the `wc_payments_disputes_summary` option key) and is invalidated (deleted) when a dispute related webhook is received. Because of this approach, there may be some delay in invalidating the cache after accepting or challenging a dispute - this will need to be discussed as part of CR. 

This PR also directs the user automatically to the disputes list table which need a response if there are any disputes which need a response. If none exist, they will just be directed to all disputes. 

#### Testing instructions

- Checkout this branch `fix/354` branch of WC Payments and `fix/client-354-m2` of WC Pay Server.
- Make sure your store is listening to WC Pay webhooks (see [this guide](https://github.com/Automattic/woocommerce-payments-server/tree/trunk/local#5-listen-to-webhooks)). We use webhook events to invalidate the cache. 
- Make a purchase using a card that will automatically raise a dispute (eg `4000 0000 0000 0259` see more [here](https://woocommerce.com/document/payments/testing/#section-15))
- Go to **Payments > Disputes** in your WP dashboard. You should now see a dispute and a badge indicating 1 dispute needs a response.
- Repeat for multiple disputes, accept disputes (lose them) etc to test the status counts are correct.

-------------------


**Post merge**
